File: Ros2_Practice/Writing_PulisherAndSubscriber/Raw_pubsub.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import String

class PubSubNode(Node):
    def __init__(self):
        super().__init__('pub_sub_node')
        
        self.publisher_ = self.create_publisher(String, 'my_topic', 10)
        self.timer = self.create_timer(1.0, self.timer_callback)
        
        self.subscription = self.create_subscription(
            String, 'my_topic', self.listener_callback, 10)
        self.subscription  

    def timer_callback(self):
        msg = String()
        msg.data = 'Data from Python Publisher'
        self.publisher_.publish(msg)

    def listener_callback(self, msg):
        self.get_logger().info(f'Received: "{msg.data}"')

def main(args=None):
    rclpy.init(args=args)
    node = PubSubNode()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: Ros2_Practice/Writing_PulisherAndSubscriber/Raw_pubsub.cpp
#include <chrono>
#include <memory>
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

using namespace std::chrono_literals;

class PubSubNode : public rclcpp::Node
{
public:
  PubSubNode() : Node("pub_sub_node")
  {
    publisher_ = this->create_publisher<std_msgs::msg::String>("my_topic", 10);
    
    timer_ = this->create_wall_timer(
      1000ms, std::bind(&PubSubNode::timer_callback, this));

    subscription_ = this->create_subscription<std_msgs::msg::String>(
      "my_topic", 10, std::bind(&PubSubNode::topic_callback, this, std::placeholders::_1));
  }

private:
  void timer_callback()
  {
    auto msg = std_msgs::msg::String();
    msg.data = "Data from C++ Publisher";
    publisher_->publish(msg);
  }

  void topic_callback(const std_msgs::msg::String::SharedPtr msg) const
  {
    RCLCPP_INFO(this->get_logger(), "Received: '%s'", msg->data.c_str());
  }

  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<PubSubNode>());
  rclcpp::shutdown();
  return 0;
}